include/mqtt/endpoint.hpp

```diff
@@ -1005,6 +1005,46 @@ class endpoint : public std::enable_shared_from_this<endpoint<Mutex, LockGuard,
         );
     }
 
+    /**
+     * @brief Publish with already acquired packet identifier
+     * @param packet_id
+     *        packet identifier. It should be acquired by acquire_unique_packet_id, or register_packet_id.
+     *        The ownership of  the packet_id moves to the library.
+     *        If qos == qos::at_most_once, packet_id must be 0. But not checked in release mode due to performance.
+     * @param topic_name
+     *        A topic name to publish
+     * @param contents
+     *        The contents to publish
+     * @param pubopts
+     *        qos, retain flag, and dup flag.
+     * @param life_keeper
+     *        An object that stays alive as long as the library holds a reference to any other parameters.
+     *        If topic_name, contents, or props do not have built-in lifetime management, (e.g. MQTT_NS::buffer)
+     *        use this parameter to manage their lifetime.
+     *
+     * @note If your QOS level is exactly_once or at_least_once, then the library will store this publish
+     *       internally until the broker has confirmed delivery, which may involve resends, and as such the
+     *       life_keeper parameter is important.
+     */
+    void publish(
+        packet_id_t packet_id,
+        as::const_buffer topic_name,
+        as::const_buffer contents,
+        publish_options pubopts = {},
+        any life_keeper = {}
```

Comment:
I'm not sure that it's a good idea to default life_keeper value for the as::const_buffer version.

Very likely end-user will be confused by this, and not provide life-keeper even when qos != 0.

Reply:
I think that it is documentation domain.

Adding document something like as follows:

```
If qos is at_least_once or exactly_once, then the lifetime of topic_name and contents 
should be kept by caller's responsibility. You can pass the object that are referred by
lifetime of topic_name and contents to the life_keeper.
```
assert(
    (pubopts.get_qos() == qos::at_most_once) ||
    ((pubopts.get_qos() == qos::at_least_once) || pubopts.get_qos() == qos::exactly_once) && life_keeper.has_value()
);
```

I also thought that the following approach:

```cpp
assert((qos == 0) || ((qos == 1) || (qos == 2) && life_keeper.has_value());
```

However, it is too limited. User can pass topic_name and contents as string literal. In this case, no life_keeper is required.

Reply:
Personally I think it's important to make it difficult to do the wrong thing.

But I don't care much about the blocking apis, so it's fine if you want to do things this way.

Reply:
I agree. I remove default parameter from `as::buffer` overload.
I don't remove `MQTT_NS::buffer` overload because typical case is `MQTT_NS::buffer` contains its lifetime.